src/timer/windows/SDL_systimer.c

```diff
@@ -146,19 +146,31 @@ SDL_GetPerformanceFrequency(void)
 void
 SDL_Delay(Uint32 ms)
 {
-    /* Sleep() is not publicly available to apps in early versions of WinRT.
-     *
-     * Visual C++ 2013 Update 4 re-introduced Sleep() for Windows 8.1 and
-     * Windows Phone 8.1.
-     *
-     * Use the compiler version to determine availability.
-     *
-     * NOTE #1: _MSC_FULL_VER == 180030723 for Visual C++ 2013 Update 3.
-     * NOTE #2: Visual C++ 2013, when compiling for Windows 8.0 and
-     *    Windows Phone 8.0, uses the Visual C++ 2012 compiler to build
-     *    apps and libraries.
-     */
-#if defined(__WINRT__) && defined(_MSC_FULL_VER) && (_MSC_FULL_VER <= 180030723)
+    /* CREATE_WAITABLE_TIMER_HIGH_RESOLUTION flag was added in Windows 10 version 1803.
+    * 
+    * Sleep() is not publicly available to apps in early versions of WinRT.
+    *
+    * Visual C++ 2013 Update 4 re-introduced Sleep() for Windows 8.1 and
+    * Windows Phone 8.1.
+    *
+    * Use the compiler version to determine availability.
+    *
+    * NOTE #1: _MSC_FULL_VER == 180030723 for Visual C++ 2013 Update 3.
+    * NOTE #2: Visual C++ 2013, when compiling for Windows 8.0 and
+    *    Windows Phone 8.0, uses the Visual C++ 2012 compiler to build
+    *    apps and libraries.
+    */
+#if NTDDI_VERSION >= NTDDI_WIN10_RS4
```

Comment:
This may need to be a runtime check. The article also notes that this functionality isn't available with the mingw toolchain. Do you know if that was a compile time issue or a runtime issue?

Maybe this should be `#ifdef CREATE_WAITABLE_TIMER_HIGH_RESOLUTION` instead?

Reply:
I agree that checking for CREATE_WAITABLE_TIMER_HIGH_RESOLUTION is better since it is only defined in Window 10 version 1803 and later anyway :
```
syncapi.h
...
#if (_WIN32_WINNT >= _NT_TARGET_VERSION_WIN10_RS4)
#define CREATE_WAITABLE_TIMER_HIGH_RESOLUTION 0x00000002
#endif
```

You're also right about the runtime check, it should fallback to other sleep methods if the timer can't be created for whatever reason. I changed it, but I still kept the compile time check because there's no point to repeatedly try to create a high resolution timer if it has no chances to succeed because the feature is not available.

Unfortunately, I have no experience with MinGW, so I have no definitive answer to your question, but I found that CREATE_WAITABLE_TIMER_HIGH_RESOLUTION was not defined in WinGW-W64 syncapi.h header, so I think it may be a compile time issue. Anyway, the code will now fallback to the previous sleep methods in both compile time (no CREATE_WAITABLE_TIMER_HIGH_RESOLUTION macro defined) and runtime (the timer can't be created) failure, therefore it should build and work fine on MinGW.

Reply:
Having `CREATE_WAITABLE_TIMER_HIGH_RESOLUTION` defined during the build doesn't necessarily mean anything about where the resulting binaries can run. We still need to load `CreateWaitableTimerEx()` and `SetWaitableTimerEx()` at runtime, since they are unavailable on XP.